compiler/rustc_data_structures/src/memmap.rs

```diff
@@ -11,33 +12,48 @@ pub struct Mmap(Vec<u8>);
 
 #[cfg(not(any(miri, target_arch = "wasm32")))]
 impl Mmap {
-    /// # Safety
-    ///
     /// The given file must not be mutated (i.e., not written, not truncated, ...) until the mapping is closed.
     ///
-    /// However in practice most callers do not ensure this, so uses of this function are likely unsound.
+    /// This process must not modify nor remove the backing file while the memory map lives.
+    /// For the dep-graph and the work product index, it is as soon as the decoding is done.
+    /// For the query result cache, the memory map is dropped in save_dep_graph before calling
+    /// save_in and trying to remove the backing file.
+    ///
+    /// There is no way to prevent another process from modifying this file.
+    ///
+    /// This means in practice all uses of this function are theoretically unsound, but also
+    /// the way rustc uses `Mmap` (reading bytes, validating them afterwards *anyway* to detect
+    /// corrupted files) avoids the actual issues this could cause.
+    ///
+    /// Someone may truncate our file, but then we'll SIGBUS, which is not great, but at least
+    /// we won't succeed with corrupted data.
+    ///
+    /// To get a bit more hardening out of this we will set the file as readonly before opening it.
```

Comment:
can't we immediately set it to readonly after initial creation? The writing process can have an FD with write permission and then make the file readonly that other processes can't open it for writing.

Reply:
Hrm, what purpose is this check supposed serve? Can we remove it?
Having intermediate files that will be mmaped remain writable has caused unsoundness before (#139453).